File: frontend/src/lib/api/scratchpad.ts
import api from './client'

// Scratchpad types
export interface ScratchpadEntry {
  id: number
  project_id: number
  numeric_id: number
  content: string
  context_hint: string | null
  resolved: boolean
  resolved_into_type: string | null
  resolved_into_id: number | null
  created_at: string
  updated_at: string
}

// API functions - Scratchpad
export const scratchpadApi = {
  list: (projectId: number, resolved?: boolean) =>
    api.get<{ entries: ScratchpadEntry[]; total: number }>(`/projects/${projectId}/scratchpad`, {
      params: { resolved },
    }).then(res => res.data),

  create: (projectId: number, data: { content: string; context_hint?: string }) =>
    api.post<ScratchpadEntry>(`/projects/${projectId}/scratchpad`, data).then(res => res.data),

  update: (projectId: number, entryId: number, data: { content?: string; resolved?: boolean }) =>
    api.patch<ScratchpadEntry>(`/projects/${projectId}/scratchpad/${entryId}`, data).then(res => res.data),

  delete: (projectId: number, entryId: number) =>
    api.delete(`/projects/${projectId}/scratchpad/${entryId}`).then(res => res.data),

  convert: (projectId: number, entryId: number, data: { target_type: string; entity_type: string; entity_id: number }) =>
    api.post<ScratchpadEntry>(`/projects/${projectId}/scratchpad/${entryId}/convert`, data).then(res => res.data),
}

File: backend/app/schemas/scratchpad.py
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from .common import UTCTimestamp


class ScratchpadEntryCreate(BaseModel):
    content: str = Field(..., min_length=1)
    context_hint: str | None = Field(None, max_length=255)


class ScratchpadEntryUpdate(BaseModel):
    content: str | None = Field(None, min_length=1)
    resolved: bool | None = None


class ScratchpadConvertRequest(BaseModel):
    target_type: str = Field(..., pattern=r'^(memo)$')
    entity_type: str = Field(..., pattern=r'^(project|conversation|document|code|code_category|analysis|dataset|dataset_row|dataset_column|canvas)$')
    entity_id: int


class ScratchpadEntryResponse(BaseModel):
    id: int
    project_id: int
    numeric_id: int
    content: str
    context_hint: str | None
    resolved: bool
    resolved_into_type: str | None
    resolved_into_id: int | None
    created_at: UTCTimestamp
    updated_at: UTCTimestamp

    model_config = ConfigDict(from_attributes=True)


class ScratchpadListResponse(BaseModel):
    entries: list[ScratchpadEntryResponse]
    total: int
